### ISSUE TITLE
[BuildQualityChecks] Failing due to timeout suddenly

### ISSUE THREAD

**Issue by timcosta** [MicrosoftPremier/VstsExtensions#159]
**Describe the context**
- **Extension:** `BuildQualityChecks`
- **Environment:** cloud
- **Agent type:** MS-hosted
- **Pipeline type:** yaml

**Describe the problem and expected behavior**
On Oct 26th at 12:15p CT we had a successful build of our pipeline using `BuildQualityChecks`. The same day at 6:58 CT all building using `BuildQualityChecks` started failing due to timeouts. We did not make any meaningful changes to our pipeline in the commit where failures started appearing, and both commits used `BuildQualityChecks@v8.1.3` so I don't believe a version update is at fault here.

The error we are experiencing is this:
```
Waiting for previous tasks to upload information...
Previous build steps did not finish within the maximum wait time.
[ERROR] Previous build steps did not finish within the maximum wait time.
##[error]Previous build steps did not finish within the maximum wait time.
```

I believe this is similar to #150, but we are already using the version mentioned in that issue. that solved the issue they were experiencing.

**Task logs**
I've emailed the task logs to PSGerExtSupport@microsoft.com and referenced this issue number in the subject.

**Comment by ReneSchumacher** [comment]
Hi @timcosta,

I'm going to look through the logs and update the issue accordingly. According to the Azure DevOps status page at https://status.dev.azure.com/_history, there have been issues on October 26th, which might have caused that behavior.

The output you see is only logged if the BQC task needs to wait for log information from previous tasks. That information is uploaded asynchronously, so it can happen that all previous tasks are marked as finished in the UI, but log information is still being uploaded in the background. If that upload is too slow, you may run into this type of issue.

There's also a variable you may use to increase the wait time when the issue occurs often. Just set the variable `PSGer.Build.MaxWaitTime` to the appropriate time in milliseconds (e.g., 600000 for 10 minutes). The default value is 4 minutes.

**Comment by ReneSchumacher** [comment]
Hi again,

the log files don't provide much insight. I can only see that the task is waiting for the "Run Tests" step to finish and then times out after the default of 4 minutes. Does the issue still persist?

**Comment by timcosta** [comment]
Hey Rene -

The issue does still persist, we haven't had a successful build since the time I noted that it started failing. I can try to increase the timeout today, but previously the task finished in about 1 second so it seems unlikely for the time to jump to over four minutes.

**Comment by ReneSchumacher** [comment]
Thanks for the update. I'd like to try the increased timeout first. There haven't been any changes to the task but there have been a couple deployments to Azure DevOps Services since October 26th.

If the issue has been introduced with one of the product deployments, I cannot do much as I am not part of the Azure DevOps product group. We could create a ticket at https://developercommunity.visualstudio.com/search?space=21 and describe the issue. Once the ticket is in place I can work with our support staff and try to figure out what might have changed.

In the log file you can see that the "Run Tests" task keeps running. Look at the `timeline` object in the `RAW DATA` section, its state is always 1, which means it is in progress. You're running a command line with that step. Can you tell me what command line tool you're running? Has anything changed around that tool?

**Comment by timcosta** [comment]
Here's the steps for the `Run Tests` task, showing order of operations and all commands being called:

```yaml
steps:
- script: |
    brew install tfenv
    tfenv install
    pip install --upgrade pip
    pip install --upgrade awscli
    export $(cat env_vars | xargs)
    npm install aws-cdk@$CDK_VERSION
    pip install -r requirements.txt
    pip install -r ingestion/tests/requirements.txt
  displayName: Install Dependencies
- script: |
    pytest --doctest-modules \
      --junitxml=junit/test-results.xml \
      --cov \
      --cov-config=.coveragerc \
      --cov-report=xml \
      ingestion/tests
  displayName: Run Tests
- task: PublishTestResults@2
  condition: succeededOrFailed()
  inputs:
    testResultsFiles: '**/test-*.xml'
    testRunTitle: Publish test results
    displayName: Publish Test Results
- task: PublishCodeCoverageResults@1
  inputs:
    codeCoverageTool: Cobertura
    summaryFileLocation: '$(System.DefaultWorkingDirectory)/**/coverage.xml'
    displayName: Publish Code Coverage
- task: BuildQualityChecks@8
  displayName: 'Check build quality'
  inputs:
    showStatistics: true
    checkCoverage: true
    coverageFailOption: 'fixed'
    coverageType: 'lines'
    coverageThreshold: '90'
```

Will run with an increased timeout now and see if that changes anything.

**Comment by timcosta** [comment]
Interesting, it passed after 5:35. curious that with no pipeline changes the time went from <5 seconds to over 5.5 minutes.

**Comment by ReneSchumacher** [comment]
Hm, that's strange. Did perhaps the pytest module change? I checked and there weren't any changes to the hosted agent VMs that might have an impact like that. Or did you perhaps enable debug logging for your pipeline before the issue occurred?

For now, I would keep the increased timeout as a workaround and watch the build duration for a couple more days. If by then the performance hasn't increased, I would probably create the support ticket I mentioned. Let me know if you do so, so I can work with the product group and give them all the internal information they need.

**Comment by ReneSchumacher** [comment]
Hi @timcosta,

I'd like to close this issue as I believe there's nothing we can do from the task side to speed up the processing of code coverage in the backend. If you are still affected by the issue, please let me know so I can get you in contact with someone from the Azure DevOps Testing Tools team for further investigation.

Unless there's additional feedback from your side, I would close the issue next week.

**Comment by timcosta** [comment]
Sounds good, let's close it. We haven't seen this again.
